### ISSUE TITLE
Custom Plugin matching a method

### ISSUE THREAD

**Issue by vanniktech** [google/error-prone#443]
I'm trying to write my own custom checks however I'm having a hard time matching a method call.

For instance I want to report an error when someone calls `MyClass.create(SomeInterface<T>)`.

Using the code below won't catch it.

```
// Ommitted the @AutoService & @BugPattern for brevity.
public class MyCustomBugChcker extends BugChecker implements MethodInvocationTreeMatcher {
    private static final Matcher<ExpressionTree> CREATE_METHOD = staticMethod().
            onClass(MyClass.class.getName())
            .named("create")
            .withParameters(MyClass.SomeInterface.class.getName());

    @Override
    public Description matchMethodInvocation(final MethodInvocationTree tree, final VisitorState state) {
        if (CREATE_METHOD.matches(tree, state)) {
            return describeMatch(tree);
        }

        return NO_MATCH;
    }
}
```

My guess is that there's a problem with generics, how can I work around this?

**Comment by vanniktech** [comment]
Decided to go for a lint check instead.

**Comment by lowasser** [comment]
Is the interface the actual parameter type of the method, or a particular type you don't want passed to the method?  You might need `argument(0, isSubTypeOf(...))`

**Comment by vanniktech** [comment]
It's the actual parameter type.

**Comment by lowasser** [comment]
That seems odd.  If we could see a repro example it'd help us figure out what's going on.

**Comment by vanniktech** [comment]
Unfortunately I don't have my sample anymore. Back at the time I was also using Error Prone 2.0.10 so it could be that this was fixed with one of the newer versions.

SomeInterface looked like this:

```java
public interface SomeInterface<T> {
  void accept(@NonNull T t) throws Exception;
}
```

and the create Method looked like this:

```java
public class MyClass<T> {
  public static MyClass<T> create(SomeInterface<T> SomeInterface) {
      ...
  }
}
```

**Comment by vanniktech** [comment]
@lowasser I actually found it today - somehow on an old stale branch:

```
@AutoService(BugChecker.class)
@BugPattern(name = "ObservableSubscribeCheck", category = JDK, summary = "Subscribing without onError", severity = ERROR, maturity = MATURE, linkType = CUSTOM, link = "example.com/bugpattern/ObservableSubscribeCheck")
public class ObservableSubscribeCheck extends BugChecker implements MethodInvocationTreeMatcher {

    Matcher<ExpressionTree> PRINT_METHOD = instanceMethod().onExactClass("rx.Observable").named("subscribe").withParameters("rx.functions.Action1");

    @Override
    public Description matchMethodInvocation(MethodInvocationTree tree, VisitorState state) {

        if (PRINT_METHOD.matches(tree, state)) {
            return describeMatch(tree);
        }

        return NO_MATCH;
    }
}
```

I also had the `MyCustomCheck` in the same module and verified that the ingration itself worked since `MyCustomCheck` flagged code (String formatting inside print method - example: `System.err.print(String.format("Hello: %s\n", Arrays.toString(new int[] { 3, 5})));`).

This was my configuration at the time:

```groovy
buildscript {
  repositories {
    maven {
      url "https://plugins.gradle.org/m2/"
    }
  }
  dependencies {
    classpath "net.ltgt.gradle:gradle-apt-plugin:0.6"
  }
}

apply plugin: 'java'

dependencies {
  compileOnly 'com.google.errorprone:error_prone_core:2.0.10-SNAPSHOT'
  compileOnly 'com.google.errorprone:error_prone_annotation:2.0.10-SNAPSHOT'
  compileOnly 'io.reactivex:rxjava:1.1.5'

  compileOnly 'com.google.auto.service:auto-service:1.0-rc2'
  apt         'com.google.auto.service:auto-service:1.0-rc2'
}
```

**Comment by ronshapiro** [comment]
@vanniktech Could it have been that you used `getName()` instead of `getCanonicalName()`?

**Comment by vanniktech** [comment]
Where exactly?

**Comment by ronshapiro** [comment]
https://github.com/google/error-prone/issues/443#issue-168319396

**Comment by vanniktech** [comment]
I actually used `Matcher<ExpressionTree> PRINT_METHOD = instanceMethod().onExactClass("rx.Observable").named("subscribe").withParameters("rx.functions.Action1");`
